File: vital-items/src/main/java/me/xra1ny/vital/items/VitalItemStackBuilder.java
package me.xra1ny.vital.items;

import lombok.Getter;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.Map.Entry;

/**
 * A builder class for creating ItemStack objects with custom attributes.
 * This class provides a fluent builder pattern for creating items with names, lore, enchantments, and more.
 *
 * @author xRa1ny
 */
@Getter(onMethod = @__(@NotNull))
public final class VitalItemStackBuilder {
    private String name;

    private Material type = Material.COBBLESTONE;

    private List<String> lore = new ArrayList<>();

    private List<ItemFlag> itemFlagList = new ArrayList<>();

    private Map<Enchantment, Integer> enchantmentLevelMap = new HashMap<>();

    private int amount = 1;

    private boolean unbreakable;

    private Map<NamespacedKey, Map.Entry<PersistentDataType<?, ?>, ?>> namespacedKeyMap = new HashMap<>();

    public VitalItemStackBuilder name(String name) {
        this.name = name;

        return this;
    }

    public VitalItemStackBuilder type(Material type) {
        this.type = type;

        return this;
    }

    public VitalItemStackBuilder lore(List<String> lore) {
        this.lore = lore;

        return this;
    }

    public VitalItemStackBuilder lore(String lore) {
        if (this.lore == null) {
            this.lore = new ArrayList<>();
        }

        this.lore.add(lore);

        return this;
    }

    public VitalItemStackBuilder itemFlags(List<ItemFlag> itemFlagList) {
        this.itemFlagList = itemFlagList;

        return this;
    }

    public VitalItemStackBuilder itemFlag(ItemFlag itemFlag) {
        if (this.itemFlagList == null) {
            this.itemFlagList = new ArrayList<>();
        }

        this.itemFlagList.add(itemFlag);

        return this;
    }

    public VitalItemStackBuilder enchantments(Map<Enchantment, Integer> enchantmentLevelMap) {
        this.enchantmentLevelMap = enchantmentLevelMap;

        return this;
    }

    public VitalItemStackBuilder enchantment(Enchantment enchantment, int enchantmentLevel) {
        if (this.enchantmentLevelMap == null) {
            this.enchantmentLevelMap = new HashMap<>();
        }

        this.enchantmentLevelMap.put(enchantment, enchantmentLevel);

        return this;
    }

    public VitalItemStackBuilder amount(int amount) {
        this.amount = amount;

        return this;
    }

    public VitalItemStackBuilder unbreakable(boolean unbreakable) {
        this.unbreakable = unbreakable;

        return this;
    }

    public <Z> VitalItemStackBuilder namespacedKey(@NotNull String key, @NotNull PersistentDataType<?, Z> persistentDataType, @NotNull Z value) {
        final NamespacedKey namespacedKey = new NamespacedKey("vital", key);

        namespacedKeyMap.put(namespacedKey, Map.entry(persistentDataType, value));

        return this;
    }

    public <Z> VitalItemStackBuilder namespacedKey(@NotNull NamespacedKey namespacedKey, @NotNull PersistentDataType<?, Z> persistentDataType, @NotNull Z value) {
        namespacedKeyMap.put(namespacedKey, Map.entry(persistentDataType, value));

        return this;
    }

    /**
     * Converts the builder's configuration into an ItemStack.
     *
     * @return The constructed ItemStack.
     */
    @SuppressWarnings("deprecation")
    @NotNull
    public <Z> ItemStack build() {
        // Create ItemStack and ItemMeta
        final ItemStack item = new ItemStack((type != null ? type : Material.COBBLESTONE), amount);

        if (this.type != Material.AIR) {
            final ItemMeta meta = item.getItemMeta();
            meta.getPersistentDataContainer().set(NamespacedKeys.ITEM_UUID, PersistentDataType.STRING, UUID.randomUUID().toString());

            if (this.name != null) {
                if (this.name.isBlank()) {
                    meta.setDisplayName(ChatColor.RESET.toString());
                } else {
                    meta.setDisplayName(this.name);
                }
            }

            // Set Enchantments if set
            if (enchantmentLevelMap != null && !enchantmentLevelMap.isEmpty()) {
                for (Entry<Enchantment, Integer> entrySet : enchantmentLevelMap.entrySet()) {
                    meta.addEnchant(entrySet.getKey(), entrySet.getValue(), true);
                }
            }

            // Set ItemFlags if set, else use all
            if (itemFlagList != null && !itemFlagList.isEmpty()) {
                for (ItemFlag itemFlag : itemFlagList) {
                    meta.addItemFlags(itemFlag);
                }
            } else {
                for (ItemFlag itemFlag : ItemFlag.values()) {
                    meta.addItemFlags(itemFlag);
                }
            }

            // Set Lore if set
            if (!lore.isEmpty()) {
                meta.setLore(lore);
            }

            meta.setUnbreakable(unbreakable);

            if(!namespacedKeyMap.isEmpty()) {
                final PersistentDataContainer persistentDataContainer = meta.getPersistentDataContainer();

                for(Map.Entry<NamespacedKey, Map.Entry<PersistentDataType<?, ?>, ?>> entry : namespacedKeyMap.entrySet()) {
                    final NamespacedKey namespacedKey = entry.getKey();
                    final PersistentDataType<?, Z> persistentDataType = (PersistentDataType<?, Z>) entry.getValue().getKey();
                    final Z value = (Z) entry.getValue().getValue();

                    if(!persistentDataType.getComplexType().equals(value.getClass())) {
                        continue;
                    }

                    persistentDataContainer.set(namespacedKey, persistentDataType, value);
                }
            }

            // Set created ItemStack's ItemMeta
            item.setItemMeta(meta);
        }

        return item;
    }
}

File: vital-items/src/main/kotlin/me/vitalframework/items/VitalItemStackBuilder.kt
package me.vitalframework.items

import net.kyori.adventure.text.format.TextDecoration
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import java.util.*

/**
 * A builder class for creating and configuring item stacks with customizable properties.
 *
 * This class provides a flexible way to define and modify item stack attributes such as type, name,
 * lore, enchantments, and other metadata. It allows for precise customization of items within the
 * Vital framework or for general Bukkit/Spigot development.
 *
 * Properties defined in this class can be adjusted to match specific item requirements, enabling
 * developers to create items with unique visual or functional characteristics.
 *
 * ### Features:
 * - Ability to set the type of the item (default is `Material.COBBLESTONE`).
 * - Option to specify a custom display name for the item.
 * - Configurable stack size through the `amount` property.
 * - Option to make the item unbreakable.
 * - Support for adding lore (descriptive text) to the item.
 * - Management of `ItemFlag` properties to customize the item's appearance and behavior.
 * - Ability to add enchantments and specify their levels.
 *
 * This class serves as a base for configuring items in a modular and readable way.
 */
open class VitalItemStackBuilder {
    var type = Material.COBBLESTONE
    var name: String? = null
    var amount = 1
    var unbreakable = false
    var lore = mutableListOf<String>()
    var itemFlags = mutableListOf<ItemFlag>()
    var enchantments = mutableMapOf<Enchantment, Int>()
}

/**
 * Constructs an `ItemStack` using the provided initialization logic
 * defined within the `VitalItemStackBuilder`.
 *
 * @param init A lambda function with a `VitalItemStackBuilder` receiver,
 * allowing customization of the item stack properties (e.g., type, name,
 * lore, enchantments, etc.).
 * @return The constructed `ItemStack` instance with the specified
 * properties applied.
 */
inline fun itemBuilder(itemUuid: UUID? = null, init: VitalItemStackBuilder.() -> Unit): ItemStack {
    val itemStackBuilder = VitalItemStackBuilder().apply { init() }

    return ItemStack(itemStackBuilder.type, itemStackBuilder.amount).apply {
        if (type == Material.AIR) return this
        // since we know we have an item which is not of type AIR, we now have a persistent data container
        itemMeta = itemMeta!!.apply {
            // each item MUST have a unique identifier, used in interactive items.
            persistentDataContainer[VitalNamespacedKey.ITEM_UUID, PersistentDataType.STRING] =
                itemUuid?.toString() ?: UUID.randomUUID().toString()

            if (itemStackBuilder.name != null) {
                displayName(
                    MiniMessage.miniMessage().deserialize("<reset><white>${itemStackBuilder.name}")
                        .decoration(TextDecoration.ITALIC, false)
                )
            }

            for ((enchantment, level) in itemStackBuilder.enchantments) {
                addEnchant(enchantment, level, true)
            }

            for (itemFlag in itemStackBuilder.itemFlags) {
                addItemFlags(itemFlag)
            }

            lore(itemStackBuilder.lore.map { MiniMessage.miniMessage().deserialize(it).decoration(TextDecoration.ITALIC, false) })

            isUnbreakable = itemStackBuilder.unbreakable
        }
    }
}